### ISSUE TITLE
Iframe Contents Have Not Fully Loaded Yet When Using API

### ISSUE THREAD

**Issue by CoryBond** [awslabs/amazon-quicksight-embedding-sdk#105]
I noticed this issue when using useEffect React Hooks and setting parameters based on a global state.

Basically I load up the UI sometimes it would crash with ``Uncaught TypeError: this.iframe.contentWindow is null` because somewhere in my code I tried to call:

```
  useEffect(() => {
    if (embedDashboardAPI && parameters)
      embedDashboardAPI.setParameters(parameters);
  }, [embedDashboardAPI, parameters]);
```

and though the embedDashboardAPI exists its iframe IS NOT READY to setParameters.

My current workaround to this (and for anyone else who might face this issue) is to use a eventListener on top of the container iframe. With that eventListener I set a state which marks the iframe as loaded before I make any API calls with this library. Here is what that looks like:

```
  const [qsLoaded, setQSLoaded] = useState<boolean>(false);

  useEffect(() => {
    const eventListenerCB = () => setQSLoaded(true);
    const { url } = options;
    if (_.isString(url) && !_.isNil(embeddingContainer)) {
      setQSLoaded(false);
      const newAPI = embedDashboard({
        ...options,
        url,
        container: embeddingContainer,
      });
      newAPI.iframe.addEventListener("load", eventListenerCB);
      setEmbedDashboardAPI(newAPI);
      // Cleanup function for useEffect
      return () => {
        newAPI.iframe.removeEventListener("load", eventListenerCB);
      };
    }
    return () => {};
  }, [options, onError, embeddingContainer]);
```

_____________

I do see this as a bug as the behavior is not properly described in the libraries README. Would it please be possible to fix so that users don't have to add these event listeners themselves? 

Thanks

**Comment by tulpun** [comment]
Thank you for reporting the issue. We are looking into it.

**Comment by bwelkin** [comment]
You may want to listen to `load` event on the embed dashboard instance first. Based on your code example above, please try something like:

```
...

const [isEmbedDashboardReady, setIsEmbedDashboardReady] = useState(false);

...

useEffect(() => {

    ...

    const newAPI = embedDashboard({
        ...options,
        url,
        container: embeddingContainer,
    });
    newAPI.on('load', () => {
        setIsEmbedDashboardReady(true);
    });

    ...

}, [options, onError, embeddingContainer]);

...

useEffect(() => {
    if (isEmbedDashboardReady && embedDashboardAPI && parameters) {
        embedDashboardAPI.setParameters(parameters);
    }
}, [isEmbedDashboardReady, embedDashboardAPI, parameters]);

...

```

**Comment by CoryBond** [comment]
Hey @bwelkin ,

Thanks for the suggestion. I did play around with the "onLoad" event callback in options... but now I am seeing odd behaviors with some of my dashboards. Basically some dashboards never call the loadCallback. It seems like the ones that take "parameters" as a prop are the ones having this behavior?

For now I am using the iframe event listener solution because that has proven to work.

Code I used to test:
```

const [qsLoaded, setQSLoaded] = useState<boolean>(false);

  useEffect(() => {
    const eventListenerCB = () => setQSLoaded(true);
    const { url } = options;
    if (_.isNil(error) && _.isString(url) && !_.isNil(embeddingContainer)) {
      setQSLoaded(false);
      // The below line clears out previous iframe. It seems sometimes qsEmbedMounter
      // doesn't do this when urls change and duplicates the iframes.
      embeddingContainer.innerHTML = "";
      const qsEmbedMounter = type === "console" ? embedSession : embedDashboard;
      const newAPI = qsEmbedMounter({
        ...options,
        url,
        loadCallback: eventListenerCB,
        container: embeddingContainer,
      });
      setEmbedDashboardAPI(newAPI);
    }

    return () => {};
  }, [error, options, type, embeddingContainer]);

  useEffect(() => {
    if (embedDashboardAPI && parameters && qsLoaded)
      embedDashboardAPI.setParameters(parameters);
  }, [embedDashboardAPI, parameters, qsLoaded]);
```

**Comment by Dhairyasarous** [comment]
I haven't been able to repro the issue with the onDashboardLoad event. Seems like your solution has unblocked you. Closing this issue. Feel free to reopen a new issue/contact AWS support if you are still facing problems.
